### ISSUE TITLE
cross-platform: use forward slash as directory separator for all platforms

### ISSUE THREAD

**Issue by gmeligio** [projen/projen#3284]
## Feature request

Currently `path.join` is used to compose paths from parts wherever is possible. This is the function in NodeJS commonly used for composing paths. In some places, where path.join doesn't fit the use case,
`path.posix.join` is being used to solve it. This issue appears only in Windows because `path.join` is platform-specific. I propose to always use `path.posix.join` that will use forward slash and is supported in all platforms.

Another example is for a project being used both in Windows and Linux. Some files are updated when changing platforms in `.projen/files.json` because of the use of `path.join`. This specific case can be solved in an MR changing the `JsonFile` path with `path.posix.join`, but it's related to the proposal.
![image](https://github.com/projen/projen/assets/22875166/eba253a9-af15-44c4-8f73-84af9ff23ab2)

We could avoid having to use first `path.join` and then change it to `path.posix.join` when it fails after changing platforms, if we use `path.posix.join` since the beginning. This will make the directory separator cross-platform whenever is used.

Windows supports both back and forward slashes as directory separator, [as seen in .NET docs](https://learn.microsoft.com/en-us/dotnet/api/system.io.path.directoryseparatorchar?view=net-8.0#examples)

The changes could be easy or hard depending on figuring out if there are breaking changes. I haven't figured out the breaking changes.

I can contribute to the implementation.

## Proposed Implementation

The change can be done by replacing "path" by "path/posix" in all `src/*/**.ts` like
```typescript
import * as path from "path/posix"
import { join } from "path/posix";
```

Another refactors can be done for DevEx, so `posix` is not imported from "path" but only the import "path/posix" is used. Then the functions will be used like `path.join` instead of `path.posix.join`.  Allowing to use a cross-platform separator with the function  `path.join` developers commonly use.

## Alternative implementation

Continue refactoring on a per use case basis, like https://github.com/projen/projen/issues/2344 and in my case will be to refactor `task-runtime.ts` and `dependencies.ts` to use "path/posix".  

Preventing the usage of "path" and require "path/posix" with some mechanism like an Eslint rule.

## Migration Breaking Changes

This _migration_ could have breaking changes whenever backslash (`\`) is used currently. I haven't figured out how functions aside from `join` are affected by using the POSIX-variant with "path/posix". I tried to collect the numbers of the path functions being used in projen codebase. Roughly this was the usage I gathered:
- join: 74
- dirname: 30
- resolve: 21
- basename: 13
- relative: 8
- isAbsolute: 3
- sep: 3
- normalize: 2
- extname: 2
- parse: 2
- format: 2
- posix.join: 20
- posix.sep: 3
- posix.relative: 1

Also, I haven't investigated how projen project types are affected by this change.

## Related Issues

https://github.com/projen/projen/issues/2344
https://github.com/projen/projen/issues/3272

**Comment by mrgrain** [comment]
This sounds reasonable. I don't develop on windows myself (I use WSL), nor do we currently have a proper test suite for it.
But I think it's safe to assume that anything that is persisted on disk should use `posix` paths.

**Comment by gmeligio** [comment]
Thanks. I'll check which paths are persisted on disk and draft a PR. I'll also investigate the eslint rule.

**Comment by vduits** [comment]
This would sort out #3272 as well. When not using WSL you end up having these indeed flip around and causing trouble in automation. Looking forward to the PR!

**Comment by gmeligio** [comment]
I submitted PR #3325 for the case of `.gitignore`. I changed the approach to normalize the path before it's being saved for .gitignore.

While experimenting, the `path/posix` import turned out to be tricky, as I was doubting while collecting information for the issue. Cases like `path.resolve` should be platform-specific, to be able to handle paths like `D:\my/folder/with\file`, where `path/posix` will error on Windows.

Also, it was hard to find the whole tree of `path` imports even only for `.gitignore`. If the path had already been constructed with Windows separators and a folder drive like `D:\`, then it could end up hitting edge cases when calling `path.posix.join` on that path when trying to use forward slashes, forcing to find the whole tree of modifications of the path from the root `path.join`, to avoid this scenario.

I'm still searching for other places where paths are persisted. Feel free to point out some cases that you have seen or think about. I can test both in Windows, WSL, and Kubuntu Linux. That's why I think the issue is still relevant and I think it could be left open even if that PR is merged.

**Comment by vduits** [comment]
Not sure if you caught this one yet, but `.projen/tasks.json` and `package.json` it will also use these types of paths at bundling [when you are using projen to generate lambda metadata and to bundle the code](https://projen.io/docs/integrations/aws/#aws-lambda-functions). 

Example for `tasks.json`:
![image](https://github.com/projen/projen/assets/14995659/3ef88b03-694e-44ce-9b37-8e6dd2331715)

Example `package.json`:
![image](https://github.com/projen/projen/assets/14995659/da65cb47-686d-4c67-b0ef-474c2cd35e84)

**Comment by vduits** [comment]
Unfortunately that only accounts for a single subdirectories, once you add a layer it'll start reverting and mixing behaviors.

Examples:
```json
        {
          "spawn": "bundle:foo/bar.lambda"
        },
        {
          "spawn": "bundle:foo\\sub-bar/sub-bar.lambda"
        },
        {
          "spawn": "bundle:lambdas\\athena-queries-step-function\\check-query-status-step/check-athena-query-status.lambda"
        },
```

**Comment by gmeligio** [comment]
@vduits could you update projen to `>= v0.80.4` and test again? I submitted [another fix](https://github.com/projen/projen/pull/3406) that was released in `v0.80.4`.

**Comment by vduits** [comment]
> @vduits could you update projen to `>= v0.80.4` and test again? I submitted [another fix](https://github.com/projen/projen/pull/3406) that was released in `v0.80.4`.

It seems that this one covers the all the ones that I can find our projects. Thanks for looking into it. 👍
